packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -0,0 +1,386 @@
+  Initializable,
+  UsingRegistry,
+  ReentrancyGuard,
+  UsingPrecompiles
+{
+  using SafeMath for uint256;
+  using SafeCast for uint256;
+
+  struct OwnershipAttestation {
+    address account;
+    uint256 issuedOn;
+    address signer;
+  }
+
+  // TODO ASv2 revisit linting issues & all solhint-disable-next-line max-line-length
+
+  // identifier -> issuer -> attestations
+  mapping(bytes32 => mapping(address => OwnershipAttestation[])) public identifierToAddresses;
+  // account -> issuer -> identifiers
+  mapping(address => mapping(address => bytes32[])) public addressToIdentifiers;
+  // signer => isRevoked
+  mapping(address => bool) public revokedSigners;
+
+  bytes32 public constant EIP712_VALIDATE_ATTESTATION_TYPEHASH = keccak256(
+    "OwnershipAttestation(bytes32 identifier,address issuer,address account,uint256 issuedOn)"
+  );
+  bytes32 public eip712DomainSeparator;
+
+  event EIP712DomainSeparatorSet(bytes32 eip712DomainSeparator);
+  event AttestationRegistered(
+    bytes32 indexed identifier,
+    address indexed issuer,
+    address indexed account,
+    uint256 issuedOn,
+    address signer
+  );
+  event AttestationDeleted(
+    bytes32 indexed identifier,
+    address indexed issuer,
+    address indexed account
+  );
+
+  /**
+   * @notice Sets initialized == true on implementation contracts
+   * @param test Set to true to skip implementation initialization
+   */
+  constructor(bool test) public Initializable(test) {}
+
+  /**
+   * @notice Used in place of the constructor to allow the contract to be upgradable via proxy.
+   * @param registryAddress The address of the registry core smart contract.
+   */
+  function initialize(address registryAddress) external initializer {
+    _transferOwnership(msg.sender);
+    setRegistry(registryAddress);
+    setEip712DomainSeparator();
+    // TODO ASv2 initialize any other variables here
+  }
+
+  /**
+   * @notice Sets the EIP712 domain separator for the Celo FederatedAttestations abstraction.
+   */
+  function setEip712DomainSeparator() public {
+    uint256 chainId;
+    assembly {
+      chainId := chainid
+    }
+
+    eip712DomainSeparator = keccak256(
+      abi.encode(
+        keccak256(
+          "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
+        ),
+        keccak256(bytes("FederatedAttestations")),
+        keccak256("1.0"),
+        chainId,
+        address(this)
+      )
+    );
+    emit EIP712DomainSeparatorSet(eip712DomainSeparator);
+  }
+
+  /**
+   * @notice Returns the storage, major, minor, and patch version of the contract.
+   * @return The storage, major, minor, and patch version of the contract.
+   */
+  function getVersionNumber() external pure returns (uint256, uint256, uint256, uint256) {
+    return (1, 1, 0, 0);
+  }
+
+  /**
+   * @notice Returns info about attestations (with unrevoked signers)
+   *   for an identifier produced by a list of issuers
+   * @param identifier Hash of the identifier
```

Comment:
I noticed here and elsewhere in this file we describe this as the _hash_ of the identifier. I am curious about that. In my view, this _is_ the identifier since it is the form that folks interact with.

Reply:
Yeah agreed that it's weird phrasing -- it's for consistency with `Attestations.sol`, so that folks know it's the same identifier being used. That said, maybe it's worth changing this and adding a comment that this is the same as the identifier in `Attestations.sol`

Reply:
Consistency is definitly important. I guess the term "identifier hash" makes sense if you consider phone number, email address, etc, to be the identifier. Definitly updating the comments in Attestations.sol and elsewhere would be good if we want to align on "identifier" referring to this value that is used on chain.